File: leetcode/407/P407.cpp
#include <queue>
#include <vector>
#include <iostream>
using namespace std;

//vector<vector<int>> *heightMap

class Solution {
public:
  struct Cell {
    int x, y;
    Cell(int x, int y): x(x), y(y) {}
  };
  
  static vector<vector<int>> *heightMap;
  
  struct Cmp {
    bool operator()(const Cell &a, const Cell &b) {
      return (*Solution::heightMap)[a.x][a.y] > (*Solution::heightMap)[b.x][b.y];
    }
  };
  
  int trapRainWater(vector<vector<int>>& heightMap) {
    this->heightMap = &heightMap;
    int n = heightMap.size(), m = heightMap[0].size();
    priority_queue<Cell, vector<Cell>, Cmp> heap;
    for (int i = 0; i < n; i++) {
      heap.emplace(i, 0);
      heap.emplace(i, m - 1);
    }
    for (int j = 1; j < m - 1; j++) {
      heap.emplace(0, j);
      heap.emplace(n - 1, j);
    }
  }
};

vector<vector<int>> *Solution::heightMap = 0;

int main() {
  vector<vector<int>> v{{1, 2, 3},{4, 5, 6}};
  Solution().trapRainWater(v);
  cout << Solution::heightMap;
  Solution::Cmp cmp();
}

File: leetcode/407/src/P407.java
import java.util.PriorityQueue;

/**
 * Created by yfy on 2/21/17.
 * 407. Trapping Rain Water II
 */

class Cell implements Comparable {
  int x, y;

  public Cell(int x, int y) {
    this.x = x;
    this.y = y;
  }

  public int h() {
    return Solution.hs[x][y];
  }

  @Override
  public int compareTo(Object o) {
    Cell c = (Cell) o;
    return Solution.hs[x][y] - Solution.hs[c.x][c.y];
  }
}

class Solution {
  public static int[][] hs;

  public int trapRainWater(int[][] heightMap) {
    hs = heightMap;
    PriorityQueue<Cell> queue = new PriorityQueue<>();
    int n = heightMap.length, m = heightMap.length;
    for (int i = 0; i < n; i++) {
      queue.add(new Cell(i, 0));
      queue.add(new Cell(i, m - 1));
    }
    for (int j = 1; j < m; j++) {
      queue.add(new Cell(0, j));
      queue.add(new Cell(n - 1, j));
    }
    int level = 0, sum = 0;
    while () {
      Cell cell = queue.poll();
      if (cell.h() < level) {
        sum += level - cell.h();
      }
    }
  }
}

public class P407 {

}
